nexus/codec.go

```diff
@@ -0,0 +1,204 @@
+package nexus
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+// A Stream is a container for a header and a reader.
+// It is used to stream inputs and outputs in the various client and server APIs.
+type Stream struct {
+	// Header that should include information on how to decode this stream.
+	// Headers constructed by the framework always have lower case keys.
+	// User provided keys are considered case-insensitive by the framework.
+	Header map[string]string
+	// Reader contains request or response data. May be nil for empty data.
+	Reader io.ReadCloser
+}
+
+// A LazyValue holds a value encoded in an underlying [Stream].
+//
+// ⚠️ When a LazyValue is returned from a client - if directly accessing the stream - it must be read it in its entirety
+// and closed to free up the associated HTTP connection. Otherwise the [LazyValue.Consume] method must be called.
+//
+// ⚠️ When a LazyValue is passed to a server handler, it must not be used after the returning from the handler method.
+type LazyValue struct {
+	codec  Codec
+	Stream *Stream
+}
+
+// Consume consumes the lazy value, decodes it from the underlying stream, and stores the result in the value pointed to by v.
+//
+//	var v int
+//	err := s.Consume(&v)
+func (s *LazyValue) Consume(v any) error {
+	defer s.Stream.Reader.Close()
+	return s.codec.Decode(s.Stream, v)
+}
+
+// Codec is used by the framework to serialize/deserialize input and output.
+// To customize serialization logic, implement this interface and provide your implementation to framework methods such
+// as [NewClient] and [NewHTTPHandler].
+// By default, the SDK supports serialization of JSONables, byte slices, and nils.
+type Codec interface {
```

Comment:
A codec IMO is a bad term for value conversion, though an acceptable term for byte transform. I appreciate the separation between the two concepts that has grown in Temporal and I think we should maintain this separation instead of falling back to a combined data-converter approach.

Reply:
This isn't a combined thing. It's just for value to content and back.

We discussed having a mapper (`NewReader`, `NewWriter`) interface that's more like what Temporal considers a payload codec.

I've seen Codec used to describe a value encoder + decoder before (https://github.com/gcanti/io-ts/blob/master/Codec.md).

I'm open to renaming though, alternatives:

- `ContentConverter`
- `ValueConverter`
- `ValueCodec`
- `Marshaler` (very Go specific but even the Go one targets bytes and not a reader/writer)
- `Serializer`

I'm tempted to just call this `Codec` and later add a more specific interface that `Codec` will embed for value serialization.
That way when we add bytes to bytes conversion, it'll just be more methods that codec can implement:

```
type ValueCodec interface {
  FromValue(...)
  ToValue(...)
}

type IOMapper interface {
  NewWriter(...)
  NewReader(...)
}

type Codec interface {
  ValueCodec
  IOMapper
}
```

Reply:
I don't think the word "codec" should be used for converting value to content and back. Some languages call this a serializer or a marshaler or a converter or something. If you don't like those for whatever reason, maybe we can find another that is in common use. I think codec is a bit of a strange term for this and I haven't really seen it used that way much in dev circles.

Reply:
I'm going with serializer, I don't love that it's unidirectional but I think it's common enough and not go specific that it's okay.

Reply:
I like converter personally, but at this point it's whatever. I'm just glad that developer nomenclature is winning over individual concerns about unidirectional or whatever.